cpp/src/arrow/compute/kernels/scalar_string.cc

```diff
@@ -1132,52 +1132,51 @@ void AddMatchSubstring(FunctionRegistry* registry) {
   {
     auto func = std::make_shared<ScalarFunction>("match_substring", Arity::Unary(),
                                                  &match_substring_doc);
-    auto exec_32 = MatchSubstring<StringType, PlainSubstringMatcher>::Exec;
-    auto exec_64 = MatchSubstring<LargeStringType, PlainSubstringMatcher>::Exec;
-    DCHECK_OK(func->AddKernel({utf8()}, boolean(), exec_32, MatchSubstringState::Init));
-    DCHECK_OK(
-        func->AddKernel({large_utf8()}, boolean(), exec_64, MatchSubstringState::Init));
+    for (const auto& ty : BaseBinaryTypes()) {
+      auto exec =
+          GenerateTypeAgnosticVarBinaryBase<MatchSubstring, PlainSubstringMatcher>(ty);
+      DCHECK_OK(func->AddKernel({ty}, boolean(), exec, MatchSubstringState::Init));
+    }
     DCHECK_OK(registry->AddFunction(std::move(func)));
   }
   {
     auto func = std::make_shared<ScalarFunction>("starts_with", Arity::Unary(),
                                                  &match_substring_doc);
-    auto exec_32 = MatchSubstring<StringType, PlainStartsWithMatcher>::Exec;
-    auto exec_64 = MatchSubstring<LargeStringType, PlainStartsWithMatcher>::Exec;
-    DCHECK_OK(func->AddKernel({utf8()}, boolean(), exec_32, MatchSubstringState::Init));
-    DCHECK_OK(
-        func->AddKernel({large_utf8()}, boolean(), exec_64, MatchSubstringState::Init));
+    for (const auto& ty : BaseBinaryTypes()) {
+      auto exec =
+          GenerateTypeAgnosticVarBinaryBase<MatchSubstring, PlainStartsWithMatcher>(ty);
+      DCHECK_OK(func->AddKernel({ty}, boolean(), exec, MatchSubstringState::Init));
+    }
```

Comment:
I think for anything which supports an ignore_case flag, it should be disabled or ignored for the regex matcher. Note RE2 assumes input is either UTF-8 or Latin1 encoded; we should also test invalid UTF-8.

Reply:
I believe we would also need to configure RE2 to assume Latin1 instead of UTF-8 on binary data as a result.

Reply:
*Note: binary = bytes, Binary = data type*

I tested RE2 with UTF8/Latin1 encodings using invalid ASCII/Latin/UTF8 and found RE2 does not supports finding/matching/regex if patterns or data are non-encoded bytes. Nevertheless, it does not triggers an error but fails to match anything.

[It seems RE2 is limited to decoding representation of binary numbers](https://github.com/google/re2/blob/main/re2/re2.h#L202).

Since RE2 is only used when `ignore_case` is enabled (and when all best-effort attempts fail in `match_like`), I propose to trigger a `NotImplemented` error when `ignore_case` is set for a Binary data type. The issue with this is that there are several string functions that only register Binary kernels and this will trigger the error even if they are valid encoded strings. To workaround this, we will require registering 4 kernels for each string function ([Large]BinaryType, [Large]StringType), so no _TypeAgnostic_ kernel execs.

Reply:
Hmm, this issue implies re2 does support binary: https://github.com/google/re2j/issues/84 (note this is for the Java bindings, but talks about the C++ library).

Reply:
I will test RE2 standalone for binary non-encoded support.

AFAIK, null bytes will always be hard to handle because it is ambiguous to identify if a null byte is part of the data or a null-terminating one. For example, consider a matching pattern like:
```c++
MatchSubstringOptions(/*pattern=*/"\xff\x00\xaa");
```
`pattern` is typed as `std::string` so `pattern.length()` will be incorrect. If `pattern` was typed as `const char*` it would require an explicit `length` which is unsafe because there is no way of validating if it matches the actual string. It seems we would need to use a non-encoding container such as an array (`uint8_t pattern[] = {...}`) to represent bytes.

Reply:
I think that for string/binary kernels we should note that embedding null bytes is not currently supported. Now, maybe I am missing something and it can be easily supported.

Reply:
Thanks.

Just to be clear: std::string can represent binary data; it can hold nulls without issue. Internally, it's a pointer and a length. However, you can run into null-termination issues if you ever use a constructor that is only a pointer without a length. I suspect that's what's happening somewhere in our stack. The `length` and `size` methods aren't calling `strlen` on the fly, they're just returning the stored length.

Reply:
I should also note, nulls are valid UTF-8 (it's codepoint zero), so in testing, nulls don't actually test for handling of non-UTF8 data (we should do something like a 3-byte codepoint that's been truncated), and we need to be able to support nulls. For instance:

```python
>>> b"\x00".decode("utf-8")
'\x00'
>>> b"\x80".decode("utf-8")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x80 in position 0: invalid start byte
```

Reply:
@lidavidm Thanks for all your feedback and I apologize for any confusion. I noticed that I have not been very clear in explaining myself, I mixed several concepts together. For instance, when I state that RE does not works for binary (even using Latin1) I meant in the case of `ignore_case = true` bc that is when RE2 is used in Arrow.

I am going to verify more carefully the `ignore_case` and `length` usage, build up more tests, and organize my findings/thoughts.